### ISSUE TITLE
AWS Dynamic Credentials Not Rotated

### ISSUE THREAD

**Issue by oliverisaac** [hashicorp/vault-secrets-operator#319]
**Describe the bug**

I have a VaultDynamicSecret which is issuing AWS credentials. On creation, VSO successfully fetches creds and creates the secret. After the credential TTL the creds are never refetched and are stuck with the initial value.

Restarting the VSO pods causes them to see the creds are expired (presumably using the .status.lastRenewalTime and .status.secretLease.leaseDuration fields) and correctly rotates them. But only once. And then it fails to rotate them after that.

**To Reproduce**

This is the VaultDynamicSecret I am deploying (I've included the status fields for reference, but obviously those are not in the actual deployed manifest):

```yaml
apiVersion: secrets.hashicorp.com/v1beta1
kind: VaultDynamicSecret
metadata:
  name: aws-creds
spec:
  destination:
    create: true
    name: aws-creds
  mount: aws
  path: creds/k8s-image-swapper
  renewalPercent: 75
  rolloutRestartTargets:
  - kind: Deployment
    name: k8s-image-swapper
  vaultAuthRef: default
status:
  lastGeneration: 1
  lastRenewalTime: 1691593888
  lastRuntimePodUID: c36f809c-446b-4d00-b089-f2d6990959b1
  secretLease:
    duration: 899
    id: aws/credsk8s-image-swapper/20lKjhw0e6MV7zMUldTQaKuP
    renewable: false
    requestID: cf67cd4a-fea4-bc99-caf0-8f408ec7baf8
  staticCredsMetaData:
    lastVaultRotation: 0
    rotationPeriod: 0
    ttl: 0
```

When the VSO pod is restarted it gives me this log message:

```
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T14:59:52Z	DEBUG	events	Not in renewal window after transitioning to a new leader/pod, lease_id=aws/creds/k8s-image-swapper/qaLmnOn90F2btwYK4WlI10sr, horizon=11m34.715898713s	{"type": "Normal", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67751556"}, "reason": "SecretLeaseRenewal"}
```

But then, after the horizon time it rotates the creds and gives a horizon of 0. (Notice how it _does_ correctly rotate the creds 1 time and restart the deployment as expected.)

```
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	DEBUG	events	Lease is not renewable, staticCreds=false, info=&v1beta1.VaultSecretLease{ID:"aws/creds/k8s-image-swapper/20lKjhw0e6MV7zMUldTQaKuP", LeaseDuration:899, Renewable:false, RequestID:"cf67cd4a-fea4-bc99-caf0-8f408ec7baf8"}	{"type": "Warning", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67755382"}, "reason": "SecretLeaseRenewal"}
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	INFO	Vault secret does not support periodic renewal/refresh via reconciliation	{"controller": "vaultdynamicsecret", "controllerGroup": "secrets.hashicorp.com", "controllerKind": "VaultDynamicSecret", "VaultDynamicSecret": {"name":"aws-creds","namespace":"k8s-image-swapper"}, "namespace": "k8s-image-swapper", "name": "aws-creds", "reconcileID": "94c14dfb-e03d-47cd-88e3-9a07339a6c83", "requeue": false, "horizon": "0s"}
vault-secrets-operator-controller-manager-fb474f8z9 manager 2023-08-09T15:11:28Z	DEBUG	events	Rollout restart triggered for {Deployment k8s-image-swapper}	{"type": "Normal", "object": {"kind":"VaultDynamicSecret","namespace":"k8s-image-swapper","name":"aws-creds","uid":"667f4fcc-1713-434c-b3dd-f7b17d9a37e9","apiVersion":"secrets.hashicorp.com/v1beta1","resourceVersion":"67755382"}, "reason": "RolloutRestartTriggered"}
```

This is my `sys/mounts/aws/tune`:

```json
{
  "default_lease_ttl": 900,
  "description": "",
  "force_no_cache": false,
  "max_lease_ttl": 43200
}
```

This is my `aws/roles/k8s-image-swapper`:

```json
{
  "credential_type": "assumed_role",
  "default_sts_ttl": 900,
  "iam_groups": [],
  "iam_tags": null,
  "max_sts_ttl": 900,
  "permissions_boundary_arn": "",
  "policy_arns": [],
  "policy_document": "",
  "role_arns": [
    "arn:aws:iam::12345:role/k8s-image-swapper"
  ],
  "user_path": ""
}
```

**Expected behavior**

The credentials should rotate every `renewalPercent * leaseDuration` or 675 seconds

**Environment**
* Kubernetes version: 
  * Distribution or cloud vendor : eks 1.24
* vault-secrets-operator version: 0.1.0

**Comment by duong-se** [comment]
I have the same issue with GCP Credentials

**Comment by rovangju** [comment]
This is still an issue for me

**Comment by rovangju** [comment]
Still ongoing. I have to rely on a cronjob for this ....

**Comment by sizgiyaev** [comment]
What is the ETA for fixing this one? 0.3.0 is released, but the issue is still there. AWS is one of the most used Vault backends and this problem is a fundamental one. Why isn't it being prioritized?

**Comment by sizgiyaev** [comment]
Thanks for the update @benashz

**Comment by cakolb** [comment]
Is there already an ETA or estimate about the version it will get implemented? Would be great to have this fixed.

**Comment by githubixx** [comment]
For us this is currently also a showstopper in implementing VSO. Would be great to see this fixed. Is there a release date for v0.4.0 with the fix included maybe?

**Comment by benashz** [comment]
We are hoping to get this fix into the 0.4.0 release. We don't have an exact release date set, yet. Stay tuned :)
